command/agent/local.go

```diff
@@ -462,33 +468,66 @@ func (l *localState) syncService(id string) error {
 	return err
 }
 
-// syncCheck is used to sync a service to the server
-func (l *localState) syncCheck(id string) error {
-	// Pull in the associated service if any
-	check := l.checks[id]
-	var service *structs.NodeService
-	if check.ServiceID != "" {
-		if serv, ok := l.services[check.ServiceID]; ok {
-			service = serv
+// syncChecks is used to sync checks to the server. If a check is associated
+// with a service and the service is out of sync, it will piggyback with the
+// sync so that it is updated as part of the same transaction.
+func (l *localState) syncChecks(checkIDs []string) error {
+	checkMap := make(map[string]structs.HealthChecks)
+
+	for _, id := range checkIDs {
+		if check, ok := l.checks[id]; ok {
+			checkMap[check.ServiceID] = append(checkMap[check.ServiceID], check)
 		}
 	}
-	req := structs.RegisterRequest{
-		Datacenter:   l.config.Datacenter,
-		Node:         l.config.NodeName,
-		Address:      l.config.AdvertiseAddr,
-		Service:      service,
-		Check:        l.checks[id],
-		WriteRequest: structs.WriteRequest{Token: l.config.ACLToken},
-	}
-	var out struct{}
-	err := l.iface.RPC("Catalog.Register", &req, &out)
-	if err == nil {
-		l.checkStatus[id] = syncStatus{inSync: true}
-		l.logger.Printf("[INFO] agent: Synced check '%s'", id)
-	} else if strings.Contains(err.Error(), permissionDenied) {
-		l.checkStatus[id] = syncStatus{inSync: true}
-		l.logger.Printf("[WARN] agent: Check '%s' registration blocked by ACLs", id)
-		return nil
+
+	for serviceID, checks := range checkMap {
+		// Create the sync request
+		req := structs.RegisterRequest{
+			Datacenter:   l.config.Datacenter,
+			Node:         l.config.NodeName,
+			Address:      l.config.AdvertiseAddr,
+			WriteRequest: structs.WriteRequest{Token: l.config.ACLToken},
+		}
+
+		// Attach the service if it should also be synced
+		if service, ok := l.services[serviceID]; ok {
+			if status, ok := l.serviceStatus[serviceID]; ok && !status.inSync {
+				req.Service = service
+			}
+		}
+
+		// Send single Check element for backwards compat with 0.4.x
+		if len(checks) == 1 {
+			req.Check = checks[0]
+		} else {
+			req.Checks = checks
+		}
+
+		// Perform the sync
+		var out struct{}
+		if err := l.iface.RPC("Catalog.Register", &req, &out); err != nil {
+			if strings.Contains(err.Error(), permissionDenied) {
+				for _, check := range checks {
+					l.checkStatus[check.CheckID] = syncStatus{inSync: true}
+					l.logger.Printf(
+						"[WARN] agent: Check '%s' registration blocked by ACLs",
```

Comment:
I think we should warn about the service registration (once) not the checks. Since Checks cannot be blocked currently.